### DISCUSSION TITLE
API for Cloudflare Functions

### DISCUSSION THREAD

**Discussion by stewones** [analogjs/analog#963] [Q&A] [answered]
I'm trying to deploy a project which makes a post call to `/v1/something`, it works well locally but when I deploy to Cloudflare pages I get 405 method not allowed. 

is analog supposed to also deploy the endpoints as functions? cause wrangler gives me the following:

```
05:18:16.153	Finished
05:18:16.153	Note: No functions dir at /functions found. Skipping.
05:18:16.154	Validating asset output directory
05:18:16.850	Deploying your site to Cloudflare's global network...
05:18:18.890	Uploading... (118/125)
05:18:19.416	Uploading... (120/125)
05:18:19.716	Uploading... (122/125)
05:18:19.855	Uploading... (125/125)
05:18:19.856	✨ Success! Uploaded 7 files (118 already uploaded) (1.26 sec)
```

**Comment by ferat** [comment]
Yes it is supposed to deploy them as functions. I actually had the same problem. I "fixed" it by copying package.json from an older project (where deployment to CF did work) to the new one. Hopefully a proper fix can be implemented soon

**Reply by brandonroberts** [reply]
Yes, CF functions should work already with the API. Did you adjust the output directory for the server?

https://analogjs.org/docs/features/deployment/providers#setting-the-output-directory

We should look at handling this internally also.

**Reply by stewones** [reply]
yeah I'm doing that

this is my vite config, the project is inside a NX workspace
![Screenshot 2024-03-24 at 17 33 50](https://github.com/analogjs/analog/assets/719763/8390dcfa-d033-4f5c-9486-c12fb8b93d3f)

the output
![Screenshot 2024-03-24 at 17 34 12](https://github.com/analogjs/analog/assets/719763/52b8c984-a562-47a7-b828-38f5b2ba8f12)

the command I'm using to build
`BUILD_PRESET=cloudflare-pages nx build home`

did I miss something? happy to add you to the repo if you want.

**Reply by brandonroberts** [reply]
What you have is correct. I'm seeing this behavior also, so it looks like a bug

**Reply by brandonroberts** [reply, marked as answer]
@stewones Figured out the issue. Cloudflare adjusted the output directory for functions. This should work

```ts
import { defineConfig } from 'vite';
import analog from '@analogjs/platform';

// https://vitejs.dev/config/
export default defineConfig(({ mode }) => ({
  /// ...other config
  plugins: [
    analog({
      nitro: {
        output: {
          dir: './dist/analog/public',
          serverDir: './dist/analog/public/_worker.js',
        },
      },
    }),
  ],
}));
```

CF docs: https://developers.cloudflare.com/pages/functions/advanced-mode/


Will update the docs

**Reply by stewones** [reply]
I'm confused, got a couple of questions:

1. my output doesn't have a `_worker.js` file, how to generate it?
2. should we still prepend `BUILD_PRESET=clouflare-pages` to the build command? tried other presets and it's still not generating the `_worker.js` file here.

otherwise, I get this:

```
 NX   Nitro entry is missing! Is "vite" preset correct?
 ```

**Reply by brandonroberts** [reply]
1. You don't generate it. As long as you set the `serverDir` like in the example above, it will generate the `_worker.js` directory for you
2. You don't need to prepend the build command unless you're building/testing locally. It will detect the Cloudflare environment when it runs the build in the Cloudflare environment.

**Reply by stewones** [reply]
WUT - what kind of witchcraft is this doing?

it's working now 😄

Thank you @brandonroberts

**Reply by stewones** [reply]
but FWIW it only works when deploying. wrangler doesn't know about the `_worker.js`, hence my initial confusion. I don't know why but it's generating a folder now.

![Screenshot 2024-03-27 at 12 34 39](https://github.com/analogjs/analog/assets/719763/de4411bf-f392-488c-8490-ea25dd1efbb5)

![Screenshot 2024-03-27 at 12 39 39](https://github.com/analogjs/analog/assets/719763/ca9d4463-9e81-43f7-be58-2244c7c763b4)

anyway, we can always test the analog API route in dev mode and just deploy it.
